### PR TITLE
Associated types for ShaderParam

### LINKED ISSUE #400: Refactor ShaderParam using associated types
We could hide the link type entirely, simplifying the interface and streamlining the batch usage.

### PR DESCRIPTION
**Pull request by sectopod**
Relies on the latest Rust changes in `TraitDef`.

Closes #400

### CODE REVIEW

File: src/render/shade.rs [hunk 1]
```diff
@@ -136,15 +140,18 @@ pub struct ParamDictionary {
     pub textures: Vec<NamedCell<TextureParam>>,
 }
 
-/// An associated link structure for `ParamDictionary` that redirects program
-/// input to the relevant dictionary cell.
+/// An associated parameter link for ParamDictionary
```
Comment (bjz) [inline src/render/shade.rs]: Could we change this comment to "Redirects program input to the relevant `ParamDictionary` cell"

**Comment by bjz** [comment]
Nice. Will this break you on alpha, @csherratt ?

**Comment by sectopod** [comment]
@bjz done! Too bad Travis is still lagging.

**Comment by bjz** [comment]
Looks good. r+ if @csherratt is happy.

**Comment by csherratt** [comment]
No problem with the change. r+

I'd let travis to be happy before this gets merged.

**Comment by bjz** [comment]
Still waiting on travis... :[

**Comment by indiv0** [comment]
@bjz travis build failed due to the std::io -> std::old_io migration not being updated for by glfw-sys, which has since been fixed.

@sectopod's branch builds for me locally with an updated gfx_gl with [PR #10](https://github.com/gfx-rs/gfx_gl/pull/10).

**Comment by kvark** [comment]
@indiv0 thanks! restarting Travis...

**Comment by kvark** [comment]
@indiv0 Travis still fails in glfw-sys. Are you sure it was fixed?

**Comment by kvark** [comment]
Well... if we don't merge, Travis is going to be failing on missing `associated_types` anyway. Hence, it doesn't make much sense to hold it. Merging...

**Comment by indiv0** [comment]
@kvark weird, it built for me locally. I'll take another look.

**Pull request merged**